enterprise/server/util/vfs_server/vfs_server.go

```diff
@@ -658,36 +663,46 @@ func (p *Server) createFile(ctx context.Context, request *vfspb.CreateRequest, p
 	return f, node, nil
 }
 
+func groupIDStringFromContext(ctx context.Context) string {
+	if c, err := claims.ClaimsFromContext(ctx); err == nil {
+		return c.GroupID
+	}
+	return interfaces.AuthAnonymousUser
+}
+
+var DownloadDeduper = singleflight.Group[string, struct{}]{}
+
 func (p *Server) openCASFile(ctx context.Context, node *fsNode, flags uint32) (*os.File, error) {
 	// If we can open the file directly from the file cache then use that.
 	if f, err := p.env.GetFileCache().Open(ctx, node.fileNode); err == nil {
 		return f, nil
 	}
 
-	// If the file is not in the file cache, download it to the backing
-	// directory and return a file handle to the backing file.
-	localFileName, err := random.RandomString(16)
+	dedupeKey := groupIDStringFromContext(ctx) + "-" + node.fileNode.GetDigest().GetHash()
+	_, _, err := DownloadDeduper.Do(ctx, dedupeKey, func(ctx context.Context) (struct{}, error) {
```

Comment:
is the struct{} returned used / necessary?

Reply:
`Do` is a generic function so I have to return something and `struct{}` is the best option to use as far as I know.